lib/main/atom/commands/refactorCode.tsx

```diff
@@ -0,0 +1,131 @@
+    if (!location) {
+      e.abortKeyBinding()
+      return
+    }
+    const selection = editor.getSelectedBufferRange()
+    if (selection.isEmpty()) {
+      e.abortKeyBinding()
+      return
+    }
+    const client = await deps.getClient(location.file)
+
+    const fileRange: protocol.FileRangeRequestArgs = {
+      file: location.file,
+      startLine: selection.start.row + 1,
+      startOffset: selection.start.column + 1,
+      endLine: selection.end.row + 1,
+      endOffset: selection.end.column + 1,
+    }
+
+    const actions = await getApplicableRefactorsActions(client, fileRange)
+
+    if (actions.length === 0) {
+      // TODO Show a "no applicable refactors here" message
+      e.abortKeyBinding()
+      return
+    }
+
+    const selectedAction = await selectListView({
+      items: actions,
+      itemTemplate: (item, ctx) => {
+        return (
+          <li>
+            <HighlightComponent
+              label={`${item.refactorDescription}: ${item.actionDescription}`}
+              query={ctx.getFilterQuery()}
+            />
+          </li>
+        )
+      },
+      itemFilterKey: "actionDescription",
+    })
+
+    if (selectedAction !== undefined) {
+      await applyRefactors(selectedAction, fileRange, client, deps, editor, e)
+
+      // TODO responseEdits could have renameFilename and renameLocation properties
+      // so we can call a rename command.
+    }
+  },
+}))
+
+async function getApplicableRefactorsActions(
+  client: TypescriptServiceClient,
+  range: protocol.FileRangeRequestArgs,
+) {
+  const responseApplicable = await client.execute("getApplicableRefactors", range)
+  if (responseApplicable.body === undefined || responseApplicable.body.length === 0) {
+    return []
+  }
+
+  const actions: RefactorAction[] = []
+  responseApplicable.body.forEach(refactor => {
+    refactor.actions.forEach(action => {
+      actions.push({
+        refactorName: refactor.name,
+        refactorDescription: refactor.description,
+        actionName: action.name,
+        actionDescription: action.description,
+        inlineable: refactor.inlineable !== undefined ? refactor.inlineable : true,
+      })
+    })
+  })
+
+  return actions
+}
+
+async function applyRefactors(
+  selectedAction: RefactorAction,
+  range: protocol.FileRangeRequestArgs,
+  client: TypescriptServiceClient,
+  deps: Dependencies,
+  editor: atom.TextEditor,
+  e: atom.CommandEvent,
+) {
+  const responseEdits = await client.execute("getEditsForRefactor", {
+    ...range,
+    refactor: selectedAction.refactorName,
+    action: selectedAction.actionName,
+  })
+
+  if (responseEdits.body === undefined) {
+    e.abortKeyBinding()
+    return
+  }
+
+  for (const edit of responseEdits.body.edits) {
+    await deps.withTypescriptBuffer(edit.fileName, async buffer => {
+      buffer.buffer.transact(() => {
+        for (const change of edit.textChanges.reverse()) {
+          editor.setTextInBufferRange(spanToRange(change), change.newText)
```

Comment:
Uh... I'm pretty sure it was supposed to be `buffer.setTextInRange`?

Reply:
Also, while at it, this (or very similar) code is used at least in three places already. Maybe it's a good idea to refactor that into a function along the lines of `async applyEdits(edits: FileCodeEdits[]): Proimse<void>`

Reply:
https://github.com/TypeStrong/atom-typescript/blob/b4fb237e1934ba1e7373dbdc1c7c3aeba0676130/lib/main/atom/codefix/codefixProvider.ts#L56-L65

That is one of the places you mention of, but it is the same `buffer.buffer.transact` sentence. In case of refactor that code away in a function, which would be the expression to keep?

Reply:
Um... I don't understand the question.

Pretty sure the code in `lib/main/atom/commands/refactorCode.tsx` *should* be pretty much the same as in `atom-typescript/lib/main/atom/codefix/codefixProvider.ts`.

`editor.setTextInBufferRange` in `lib/main/atom/commands/refactorCode.tsx` looks like an error to me.

Reply:
Ohh. I see, sorry, my bad. I was looking all the time at line 59. You're totally right.